File: ios/Runner/QuranTranslation/TranslationAssetSync.swift
import Foundation

/// Quran translation `AIAssetPlugin` (ADR-009). Registered lazily per catalog
/// `packId`; multiple languages coexist on disk (en, ur, fr, …).
final class TranslationAssetSync: AIAssetPlugin {
  let assetId: String
  let languageCode: String
  let kind = TranslationAssetDistributionConfig.translationKind

  init(assetId: String, languageCode: String) {
    self.assetId = assetId
    self.languageCode = languageCode
  }

  var workingDirectory: URL {
    TranslationStore.shared.languageRoot(languageCode)
  }

  func isAvailable() -> Bool {
    TranslationStore.shared.isAvailable(languageCode)
  }

  func fileSpecs(manifest: [String: Any], artifactBase: URL) throws -> [AssetDownloadManager.FileSpec] {
    guard let name = manifest["translation"] as? String else {
      throw AssetDownloadError.invalidManifest("Invalid translation manifest: missing \"translation\" key.")
    }
    return [
      AssetDownloadManager.FileSpec(
        url: artifactBase.appendingPathComponent(name),
        relativePath: name
      ),
    ]
  }

  func install(
    stagingDirectory: URL,
    manifestJSON: [String: Any],
    progress: ((Double) -> Void)? = nil
  ) throws {
    try TranslationStore.shared.installFromDirectory(
      stagingDirectory,
      languageCode: languageCode,
      manifest: manifestJSON
    )
    progress?(1.0)
  }
}

File: android/app/src/main/kotlin/com/app/deenly/deenly/qurantranslation/TranslationAssetSync.kt
package com.rnr.deenfocus.qurantranslation

import com.rnr.deenfocus.assetdownload.AIAssetPlugin
import com.rnr.deenfocus.assetdownload.AssetDownloadException
import com.rnr.deenfocus.assetdownload.AssetDownloadManager
import org.json.JSONObject

/** Quran translation [AIAssetPlugin] — one instance per catalog `packId`. */
class TranslationAssetSync(
    private val store: TranslationStore,
    override val assetId: String,
    private val languageCode: String,
) : AIAssetPlugin {
    override val kind: String = TranslationAssetDistributionConfig.TRANSLATION_KIND
    override val workingDirectory get() = store.languageRoot(languageCode)

    override fun isAvailable(): Boolean = store.isAvailable(languageCode)

    override fun fileSpecs(manifest: JSONObject, artifactBase: String): List<AssetDownloadManager.FileSpec> {
        val name = manifest.optString("translation", "")
        if (name.isEmpty()) {
            throw AssetDownloadException.InvalidManifest("Invalid translation manifest: missing \"translation\" key.")
        }
        return listOf(AssetDownloadManager.FileSpec(url = artifactBase + name, relativePath = name))
    }

    override fun install(stagingDir: java.io.File, manifestJson: JSONObject, progress: ((Double) -> Unit)?) {
        store.installFromDirectory(stagingDir, languageCode, manifestJson)
        progress?.invoke(1.0)
    }
}
